### PR TITLE
Raised WARNINGS if input arguments dont have same shape 

### PR DESCRIPTION
**Pull request by iamharsha1999**
Related to Issue #190. I have included warning checkpoints only to those functions that doesn't raise error by default in order to avoid unnecessary warnings.

### CODE REVIEW

**Comment by iamharsha1999** [comment]
@googlebot I signed it!

**Comment by iamharsha1999** [comment]
I have rectified the errors from linting and unit tests. Kindly verify and the merge.

**Comment by iamharsha1999** [comment]
Can I know why its taking so long for workflow approval ?

**Comment by iamharsha1999** [comment]
@david-berthelot @AlexeyKurakin Can you please approve the workflow and run the tests ?

File: objax/functional/loss.py [hunk 1]
```diff
@@ -38,6 +39,9 @@ def cross_entropy_logits(logits: JaxArray, labels: JaxArray) -> JaxArray:
     Returns:
         (batch, ...) tensor of the cross-entropies for each entry.
     """
+    if logits.shape != labels.shape:
+        warnings.warn('Size Mismatch: logits {} and labels {}'.format(logits.shape, labels.shape))
```
Comment (AlexeyKurakin) [inline objax/functional/loss.py:43]: Could you please add function name into each warning?
Like following: `'Argument shape mismatch in cross_entropy_logits: logits {} and labels {}'.format(logits.shape, labels.shape)'`

**Review by AlexeyKurakin — COMMENTED** [review]
Sorry for delayed review and thanks for contributing.

The change looks good overall. I would just ask you to add function name into each of the warning. I provided example in the code comments. This would help for user to actually see where warning is happening, instead of guessing.

File: objax/functional/loss.py (hunk 1 above)
Comment (iamharsha1999) [reply objax/functional/loss.py]: Hey @AlexeyKurakin have made the following changes as per your comment. I added a separate decorator for this purpose in order to keep the code simple. Kindly review the changes and merge.

Comment (AlexeyKurakin) [reply objax/functional/loss.py]: I would actually ask to use simple warnings instead of decorator. Generally decorators in Python adds a level of indirection, thus understanding the code will require looking at the code of the decorator.
In Objax we prefer to have explicit code whenever possible. This makes it easier to know what code is doing and easier to debug code.  As far I as remember, the only place where we use and encourage decorator is `objax.Function.with_vars` or `objax.Function.auto_vars`.

Also I would ask to avoid overriding warnings formatter. User of the library may have their own customization of warning formatter, which will interfere with overrides done in Objax.

Comment (iamharsha1999) [reply objax/functional/loss.py]: Ohh then would change it to the old way. No issues.

File: objax/functional/loss.py [hunk 2]
```diff
@@ -38,6 +40,8 @@ def cross_entropy_logits(logits: JaxArray, labels: JaxArray) -> JaxArray:
     Returns:
         (batch, ...) tensor of the cross-entropies for each entry.
     """
+    if logits.shape != labels.shape:
+        warnings.warn(' {} {} : arg1 {} and arg2 {}'.format(warn_mess, __all__[0], logits.shape, labels.shape))
```
Comment (AlexeyKurakin) [inline objax/functional/loss.py:44]: Using `__all__[index]` is very brittle to refer to function name.
If somebody later change order of functions in `__all__` then all warnings will be broken.
The simple way is to just explicitly write function name as a string: `warningns.warn(' .... '. format(WARN_MESS, 'cross_entropy_logits', logits.shape, labels.shape)`

You can look into programmatically getting function name from within a function.
The best option I found is `inspect.currentframe().f_code.co_name`, but it's not particularly nice or readable.
See here for some discussion: https://stackoverflow.com/questions/5067604/determine-function-name-from-within-that-function-without-using-traceback

File: objax/functional/loss.py [hunk 3]
```diff
@@ -21,11 +21,13 @@
...
 from typing import Union, Iterable, Optional
-
 import jax.numpy as jn
-
 from objax.functional.core import logsumexp
 from objax.typing import JaxArray
+import warnings
+
+
+warn_mess = 'Argument shape mismatch in'
```
[the remark this answers is not in the document]
Comment (iamharsha1999) [reply objax/functional/loss.py]: The main reason I avoided to use the function names explicitly was to keep the code length lesser than 120 characters per line. Otherwise the linter returns an error. Now I have added an another constant `FUNC_NAME` in each functions definition and then passed it to the warning message. I guess this one should be perfect. Kindly do verify.

[the remark this answers is not in the document]
Comment (AlexeyKurakin) [reply objax/functional/loss.py]: Yes, that's fine.
FYI, if line is too long you can split it into multiple lines.

**Review by AlexeyKurakin — APPROVED** [review]
